File: bench/python/errors-deep.py
import os, sys
n = int(os.environ.get("BENCH_N", "50000"))
md = 1000000007
DEPTH = 50
sys.setrecursionlimit(10000)

class BenchError(Exception):
    pass

def descend(d, i):
    if d == 0:
        raise BenchError(i % 100)
    return 1 + descend(d - 1, i)

acc = 0
for i in range(n):
    try:
        descend(DEPTH, i)
    except BenchError as e:
        acc += e.args[0]
print(acc % md)

File: bench/ruby/errors-deep.rb
n = (ENV["BENCH_N"] || "50000").to_i
md = 1000000007
DEPTH = 50

class BenchError < StandardError
  attr_reader :v
  def initialize(v); @v = v; end
end

def descend(d, i)
  raise BenchError.new(i % 100) if d == 0
  1 + descend(d - 1, i)
end

acc = 0
(0...n).each do |i|
  begin
    descend(DEPTH, i)
  rescue BenchError => e
    acc += e.v
  end
end
puts acc % md
